src/activity_handlers.cpp

```diff
@@ -2728,6 +2729,44 @@ void activity_handlers::try_sleep_finish( player_activity *act, player *p )
     act->set_to_null();
 }
 
+void activity_handlers::build_do_turn( player_activity *act, player *p )
+{
+
+    const std::vector<construction> &list_constructions = get_constructions();
+    const construction &built = list_constructions[act->index];
+    partial_con *pc = g->m.partial_con_at( act->placement );
+    if( pc == nullptr ) {
+        debugmsg( "No partial construction found at activity placement, aborting activity" );
+        g->m.remove_trap( act->placement );
+        p->cancel_activity();
+        return;
+    }
+    // item_counter represents the percent progress relative to the base batch time
+    // stored precise to 5 decimal places ( e.g. 67.32 percent would be stored as 6732000 )
+    const int old_counter = pc->counter;
+
+    // Base moves for construction with no speed modifier or assistants
+    // Must ensure >= 1 so we don't divide by 0;
+    const double base_total_moves = std::max( 1, built.time );
+    // Current expected total moves, includes construction speed modifiers and assistants
+    const double cur_total_moves = std::max( 1, built.adjusted_time() );
+    // Delta progress in moves adjusted for current crafting speed
+    const double delta_progress = p->get_moves() * base_total_moves / cur_total_moves;
+    // Current progress in moves
+    const double current_progress = old_counter * base_total_moves / 10000000.0 +
+                                    delta_progress;
+    // Current progress as a percent of base_total_moves to 2 decimal places
+    pc->counter = round( current_progress / base_total_moves * 10000000.0 );
+    p->set_moves( 0 );
+
+    pc->counter = std::min( static_cast<int>( pc->counter ), 10000000 );
```

Comment:
Don't think you need this cast, `pc->counter` is an int.
```suggestion
    pc->counter = std::min( pc->counter, 10000000 );
```